src/sdc11073/roles/waveformprovider/waveformproviderimpl.py

```diff
@@ -0,0 +1,293 @@
+        self._worker_thread.start()
+
+    def stop(self):
+        """Stop worker thread."""
+        if self._worker_thread is not None:
+            self._stop_worker.set()
+            self._worker_thread.join()
+            self._worker_thread = None
+
+    @property
+    def is_running(self) -> bool:
+        """Return True if worker thread is set."""
+        return self._worker_thread is not None
+
+    def set_activation_state(self, descriptor_handle: str, component_activation_state: ComponentActivation):
+        """Set the activation state of waveform generator and of Metric state in mdib."""
+        wf_generator = self._waveform_generators[descriptor_handle]
+        wf_generator.set_activation_state(component_activation_state)
+        with self._mdib.transaction_manager() as mgr:
+            state = mgr.get_state(descriptor_handle)
+            state.ActivationState = component_activation_state
+            # if the generator is not active, there shall be no MetricValue
+            if not wf_generator.is_active:
+                state.MetricValue = None
+
+    def update_all_realtime_samples(self, transaction: RtDataMdibUpdateTransaction):
+        """Update all realtime sample states that have a waveform generator registered.
+
+        On transaction commit the mdib will call the appropriate send method of the sdc device.
+        """
+        for descriptor_handle, wf_generator in self._waveform_generators.items():
+            if wf_generator.is_active:
+                state = transaction.get_real_time_sample_array_metric_state(descriptor_handle)
+                self._update_rt_samples(state)
+        self._add_all_annotations()
+
+    def provide_waveforms(self,
+                          generator_class: type[WaveformGeneratorProtocol] = waveforms.TriangleGenerator,
+                          max_waveforms: int | None = None) -> list[str]:
+        """Create waveform generators for waveforms in mdib.
+
+        The values of descriptor.TechnicalRange and descriptor.SamplePeriod are used for the generator.
+
+        :param generator_class: the generator to be instantiated.
+        :param max_waveforms: limit number of waveforms.
+        :return: list of handles of created generators
+        """
+        name = self._mdib.data_model.pm_names.RealTimeSampleArrayMetricDescriptor
+        all_waveforms = self._mdib.descriptions.NODETYPE.get(name)
+        if max_waveforms:
+            all_waveforms = all_waveforms[:max_waveforms]
+        for waveform in all_waveforms:
+            min_value = 0
+            max_value = 1
+            try:
+                tech_range = waveform.TechnicalRange[0]
+            except IndexError:
+                pass
+            else:
+                if tech_range.Lower is not None:
+                    min_value = float(tech_range.Lower)
+                if tech_range.Upper is not None:
+                    max_value = float(tech_range.Upper)
+                if max_value < min_value:

```

Comment:
shouldnt this be an error?

Reply:
not sure. This data comes from a mdib file and can have swapped values, may be there are reasons to do that. The waveform generator needs the values sorted, therefore they get sorted here.